### PR TITLE
billing - check whether access control is ON

### PR DESCRIPTION
**Pull request by ClaudeROSSI**
A lot of unexpected problems with that one.
This is a DRAFT - with a lot of stuff to clarify (marked with TODO comments)
@LucasBrazi06 : As discussed ... I need your advise for this PR

### CODE REVIEW

File: src/server/ocpp/utils/OCPPUtils.ts [hunk 1]
```diff
@@ -300,7 +300,7 @@ export default class OCPPUtils {
     }
   }
 
-  public static async processTransactionBilling(tenant: Tenant, transaction: Transaction, action: TransactionAction): Promise<void> {
+  public static async processTransactionBilling(tenant: Tenant, transaction: Transaction, chargingStation: ChargingStation, action: TransactionAction): Promise<void> {
```
Comment (ClaudeROSSI) [inline src/server/ocpp/utils/OCPPUtils.ts:303]: @LucasBrazi06 : As discussed I have tried to align the logic with the processTransactionRoaming approach. 
But I couldn't apply exactly the same logic and I'm not satisfied with the current solution.
I do not like the fact that the signature of the **processTransactionBilling** method expects parameters that are most of the time not necessary. The charging station information is now provided but only used during the START transaction.
From my perspective, to be on the safe side, we should avoid the **switch** on the **TransactionAction**. Each event (START, STOP, END) should have its own signature to enforce the calling layer to only pass the required information.
Let me know what you think.

Comment (LucasBrazi06) [reply src/server/ocpp/utils/OCPPUtils.ts:303]: We should have the same approach when processing this kind of component whatever you choose.
Make me a proposal.

File: src/integration/billing/stripe/StripeBillingIntegration.ts [hunk 2]
```diff
@@ -795,6 +795,12 @@ export default class StripeBillingIntegration extends BillingIntegration {
         withBillingActive: false
       };
     }
+    if (!transaction.chargeBox?.siteArea?.accessControl) {
```
[the remark this answers is not in the document]
Comment (ClaudeROSSI) [reply src/integration/billing/stripe/StripeBillingIntegration.ts]: transaction.siteArea is not set when calling start transaction. 
I now pass the information as an additional parameter.

File: src/server/ocpp/utils/OCPPUtils.ts (hunk 1 above)
Comment (ClaudeROSSI) [reply src/server/ocpp/utils/OCPPUtils.ts:303]: Well ... I would be to move the split the code that we have in OCPPUtils.processTransactionBilling and move it as static methods of the BillingIntegration class.
- BillingIntegration.processStartTransation()
- BillingIntegration.processStopTransation()
- etc

But I'm not sure the same logic will match the processTransactionRoaming() use-case.

Can you review the current solution and let me know if the current implementation is acceptable?

File: src/assets/server/rest/v1/schemas/user/user-default-tag-car-get.json [hunk 3]
```diff
@@ -6,6 +6,9 @@
     "UserID": {
       "$ref": "user#/definitions/id"
     },
+    "ChargingStationID": {
+      "$ref": "common#/definitions/string-ids"
```
Comment (LucasBrazi06) [inline src/assets/server/rest/v1/schemas/user/user-default-tag-car-get.json:10]: It's only one charging station ID so use 'string-id'

File: src/integration/billing/BillingIntegration.ts [hunk 4]
```diff
@@ -228,7 +229,7 @@ export default abstract class BillingIntegration {
     }
   }
 
-  public checkStartTransaction(transaction: Transaction): void {
+  public checkStartTransaction(transaction: Transaction, chargingStation: ChargingStation): void {
```
Comment (LucasBrazi06) [inline src/integration/billing/BillingIntegration.ts:232]: Pass also User and SiteArea in param?

File: src/integration/billing/BillingIntegration.ts [hunk 5]
```diff
@@ -247,18 +248,23 @@ export default abstract class BillingIntegration {
...
+    if (!chargingStation) {
+      throw new BackendError({
+        message: 'The charging station is mandatory to start a transaction',
+        module: MODULE_NAME,
+        method: 'checkStartTransaction',
+        action: ServerAction.BILLING_TRANSACTION
+      });
+    }
+    // Check for the Site Area
+    if (!chargingStation.siteArea) {
```
Comment (LucasBrazi06) [inline src/integration/billing/BillingIntegration.ts:260]: If Org is not active, no Site Area will be provide.
Check if the org component is active.
Missing unit tests? ;-)

File: src/server/ocpp/utils/OCPPUtils.ts (hunk 1 above)
Comment (LucasBrazi06) [reply src/server/ocpp/utils/OCPPUtils.ts:303]: I checked and pass also Site Area and User in addition to Charging Station.
I can do the same for the Roaming.

File: src/server/rest/v1/router/api/UserRouter.ts [hunk 6]
```diff
@@ -64,7 +64,11 @@ export default class UserRouter {
 
   private buildRouteUserDefaultCarTag(): void {
     this.router.get(`/${ServerRoute.REST_USER_DEFAULT_TAG_CAR}`, async (req: Request, res: Response, next: NextFunction) => {
-      req.query.ID = req.params.id;
```
Comment (LucasBrazi06) [inline src/server/rest/v1/router/api/UserRouter.ts:67]: Do not remove req.query.ID assignment as it has to work in the future when www and app will be fixed.

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

File: src/assets/server/rest/v1/schemas/user/user-default-tag-car-get.json (hunk 3 above)
Comment (ClaudeROSSI) [reply src/assets/server/rest/v1/schemas/user/user-default-tag-car-get.json]: done

File: src/server/ocpp/utils/OCPPUtils.ts (hunk 1 above)
Comment (ClaudeROSSI) [reply src/server/ocpp/utils/OCPPUtils.ts]: Same comment: 4 distinct methods in **BillingFacade** class

File: src/integration/billing/BillingIntegration.ts (hunk 4 above)
Comment (ClaudeROSSI) [reply src/integration/billing/BillingIntegration.ts]: Same comment: 4 distinct methods in **BillingFacade** class

File: src/integration/billing/BillingIntegration.ts (hunk 5 above)
Comment (ClaudeROSSI) [reply src/integration/billing/BillingIntegration.ts]: done.
I still have missing tests - but couldn't implement them with the new Jest framework!

File: src/integration/billing/BillingFacade.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  public static async processStartTransaction(tenant: Tenant, transaction: Transaction, chargingStation: ChargingStation, siteArea: SiteArea): Promise<void> {
+    if (!transaction.user || !transaction.user.issuer) {
+      return;
+    }
+    const billingImpl = await BillingFactory.getBillingImpl(tenant);
+    if (billingImpl) {
+      try {
+        let withBillingActive = false;
+        // Check parameters' consistency
+        if (billingImpl.checkStartTransaction(transaction, chargingStation, siteArea)) {
```
Comment (ClaudeROSSI) [inline src/integration/billing/BillingFacade.ts:25]: The **checkTransactionMethod**() is now explicitly called by the abstraction layer. The concrete implementation (i.e. STRIPE) does not need to take care of this logic anymore.

File: src/integration/billing/BillingIntegration.ts [hunk 8]
```diff
@@ -247,18 +251,64 @@ export default abstract class BillingIntegration {
...
 
-  private async _getUsersWithNoBillingData(): Promise<User[]> {
-    const newUsers = await UserStorage.getUsers(this.tenant,
-      {
-        statuses: [UserStatus.ACTIVE],
-        notSynchronizedBillingData: true
-      }, Constants.DB_PARAMS_MAX_LIMIT);
-    if (newUsers.count > 0) {
-      return newUsers.result;
+  protected async updateTransactionsBillingData(billingInvoice: BillingInvoice): Promise<void> {
```
Comment (ClaudeROSSI) [inline src/integration/billing/BillingIntegration.ts:283]: This code has be moved out of the STRIPE implementation. 
It is now part of the billing abstraction to make sure all future implementations can benefit from it.

File: src/server/rest/v1/service/UserService.ts [hunk 9]
```diff
@@ -81,9 +82,22 @@ export default class UserService {
         );
       }
     }
+    let withBillingChecks = true ;
+    if (filteredRequest.ChargingStationID) {
+      // TODO - The ChargingStationID is optional but only for backward compatibility reasons - make it mandatory as soon as possible
+      const chargingStation = await ChargingStationStorage.getChargingStation(req.tenant, filteredRequest.ChargingStationID,
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:88]: To use the Auth check method:
```
    const chargingStation = await UtilsService.checkAndGetChargingStationAuthorization(
      req.tenant, req.user, filteredRequest.ID, action, null, {
        withSite: true,
        withSiteArea: true,
      }, true);

```

File: src/server/rest/v1/router/api/UserRouter.ts (hunk 6 above)
Comment (LucasBrazi06) [reply src/server/rest/v1/router/api/UserRouter.ts:67]: To not forget to change the WWW et App to use this URL path correctly

**Review by LucasBrazi06 — APPROVED** [review]

**Pull request merged**